File: partiql-ast/src/main/java/org/partiql/ast/v1/Column.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;
import org.partiql.ast.v1.type.Type;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
public class Column extends AstNode {
    @NotNull
    public String name;

    @NotNull
    public Type type;

    @NotNull
    public List<Constraint> constraints;

    public Column(@NotNull String name, @NotNull Type type, @NotNull List<Constraint> constraints) {
        this.name = name;
        this.type = type;
        this.constraints = constraints;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        List<AstNode> kids = new ArrayList<>();
        kids.add(type);
        kids.addAll(constraints);
        return kids;
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitColumn(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/Column.kt
package org.partiql.ast.v1

import org.partiql.ast.v1.type.Type

/**
 * TODO docs, equals, hashcode
 */
public class Column(
    @JvmField
    public var name: String,
    @JvmField
    public var type: Type,
    @JvmField
    public var constraints: List<Constraint>,
) : AstNode() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.add(type)
        kids.addAll(constraints)
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitColumn(this, ctx)
}
